### PR TITLE
Opt In Overwrite

### LINKED ISSUE #5: [Contract] Overwrite Protection
Measure 1: Return the old key from the method in case it would be overwriting something. This should already be the behavior for the data structures provided by `near-sdk`

Measure 2: Make overwriting opt-in via an `overwrite: bool` argument

_Originally posted by @tifrel in https://github.com/Mintbase/neareth/pull/3#discussion_r1471203376_

### PR DESCRIPTION
**Pull request by bh2smith**
To protect users from overwriting their key, we have a boolean-flag (overwrite) which must be set to true to overwrite values.

Closes #5.

Adjusted Rust Tests.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -21,16 +20,19 @@ impl Default for EthKeys {
...
 impl EthKeys {
     // Sets the encrypted key for the sender's account.
-    pub fn set_key(&mut self, encrypted_key: String) {
+    pub fn set_key(&mut self, encrypted_key: String, overwrite: bool) {
         // TODO - would be nice if there was some way to validate
         // that the encrypted key actualy contains expected data.
         let account_id = env::signer_account_id();
-        self.evm_key.insert(&account_id, &encrypted_key);
-        log_str(&format!("SetKey({account_id})"));
+        if !self.evm_key.contains_key(&account_id) || overwrite {
```
Comment (tifrel) [inline src/lib.rs]: Multiline suggestions seem not to work here, so here is my suggestion:

```rs
match self.evm_key.insert(&account_id, &encrypted_key) {
  Some(old_encrypted_key) = > {
    if !overwrite {
      env::panic_str("Key already set!");
    }
    log_str(&format!("SetKey({account_id}, {encyrpted_key})"));
    Some(old_encrypted_key)
  }
  None => {
    log_str(&format!("SetKey({account_id}, {encyrpted_key})"));
    None
  },
}
```

This way we return the old key on overwrite, so the user can do with it whatever they need. By panicking, we signal to the user (and wallets) that their action failed.

File: src/lib.rs [hunk 2]
```diff
@@ -21,16 +20,19 @@ impl Default for EthKeys {
...
 
-// Implement the contract structure
 #[near_bindgen]
 impl EthKeys {
     // Sets the encrypted key for the sender's account.
-    pub fn set_key(&mut self, encrypted_key: String) {
+    pub fn set_key(&mut self, encrypted_key: String, overwrite: bool) {
         // TODO - would be nice if there was some way to validate
         // that the encrypted key actualy contains expected data.
         let account_id = env::signer_account_id();
```
Comment (tifrel) [inline src/lib.rs:34]: There are nuances around `signer_account_id` and `predecessor_account_id`, I would have to verify, but this way a rogue smart contract might be able to set a new key for a user. It's also common practice to require a yoctoNEAR deposit, which can only be done using FullAccessKeys, but not with FunctionCallKeys.

File: ts-lib/tests/local.test.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+  it("get_key", async () => {
+    const key = await contract.view("get_key", { account_id: user.accountId });
+    expect(key).toBe(null);
+  });
+  it("set_key", async () => {
+    const value = "Howdy";
+    await user.call(contract, "set_key", {
+      encrypted_key: value,
+      overwrite: false,
```
Comment (tifrel) [inline ts-lib/tests/local.test.ts]: I would consider making this optional as it's gonna be false in most cases.

**Review by tifrel — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (bh2smith) [reply src/lib.rs:34]: Very interesting. Thanks. I was under the impression that the yoco deposit would be a separate change that would also have to make the function payable. I had captured this in issue https://github.com/Mintbase/neareth/issues/2. However, your comment seems to mean two different things here (something about how other accounts can overwrite other's keys)... Are these two different things, or does having attached deposit somehow prevent this?

File: src/lib.rs (hunk 1 above)
Comment (bh2smith) [reply src/lib.rs]: I really like this suggestion, will do. Fixed in 18b8769

File: ts-lib/tests/local.test.ts (hunk 3 above)
Comment (bh2smith) [reply ts-lib/tests/local.test.ts]: Do you mean to make the Contract's input parameter optional or the ts-library functions? I had never considered making optional input parameters in a smart contract can look into it.

File: src/lib.rs (hunk 2 above)
Comment (bh2smith) [reply src/lib.rs:34]: Fixed in 726d486

File: ts-lib/tests/local.test.ts (hunk 3 above)
Comment (bh2smith) [reply ts-lib/tests/local.test.ts]: Fixed in 726d486

File: src/lib.rs (hunk 2 above)
Comment (tifrel) [reply src/lib.rs:34]: #2 covers everything, except the `signer`/`predecessor`. Most of this comment is just information about NEAR smart contracts and common practices.

Comment (bh2smith) [reply src/lib.rs:34]: Sounds a bit like the analogue of sender vs something else when contracts make delegate calls to others. So then would you say it's "approved"?

Comment (tifrel) [reply src/lib.rs:34]: Left a comment on #2, I think it's better addressed there.

**Review by tifrel — APPROVED** [review]

**Pull request merged**